handler.go

```diff
@@ -87,29 +97,47 @@ func checkHandleType(t reflect.Type) (*requestType, error) {
 
 func handleParams(fv reflect.Value, pt *requestType) func(w http.ResponseWriter, req *http.Request, p httprouter.Params) ([]reflect.Value, error) {
 	ft := fv.Type()
-	argStructType := ft.In(1).Elem()
 	returnJSON := ft.NumOut() > 1
+	needsParams := ft.In(0) == paramsType
+	if needsParams {
+		argStructType := ft.In(1).Elem()
+		return func(w http.ResponseWriter, req *http.Request, p httprouter.Params) ([]reflect.Value, error) {
+			if err := req.ParseForm(); err != nil {
+				return nil, errgo.WithCausef(err, ErrUnmarshal, "cannot parse HTTP request form")
+			}
+			if returnJSON {
+				w = headerOnlyResponseWriter{w.Header()}
+			}
+			params := Params{
+				Response: w,
+				Request:  req,
+				PathVar:  p,
+			}
+			pv := reflect.New(argStructType)
+			if err := unmarshal(params, pv, pt); err != nil {
+				return nil, errgo.NoteMask(err, "cannot unmarshal parameters", errgo.Is(ErrUnmarshal))
+			}
+			return fv.Call([]reflect.Value{
```

Comment:
fv must be a function type because we check all aspects of its type in checkHandleType (we pass fv.Type() to checkHandleType on line 58). checkHandleType is responsible for all the checking we need to do - I don't think there's any need to do the checking redundantly here.